v2/health.go

```diff
@@ -0,0 +1,127 @@
+	if options == nil {
+		options = &HealthCheckerOptions{}
+	}
+	if options.HealthCheckTimeout == 0 || options.HealthCheckTimeout > interval {
+		options.HealthCheckTimeout = interval
+	}
+
+	return &HealthChecker{
+		clients:      clients,
+		entity:       entity,
+		subscription: subscription,
+		interval:     interval,
+		options:      options,
+	}
+}
+
+// StartSenderPeriodicHealthCheck starts a periodic health check for the sender.
+// It uses azservicebus.Sender.NewMessageBatch() to perform the health check.
+// Stops when the context is cancelled.
+func (h *HealthChecker) StartSenderPeriodicHealthCheck(ctx context.Context) {
+	for namespace, client := range h.clients {
+		go h.periodicHealthCheck(ctx, h.senderHealthCheck, namespace, client)
+	}
+}
+
+// StartReceiverPeriodicHealthCheck starts a periodic health check for the receiver
+// It uses azservicebus.Receiver.PeekMessages() to perform the health check.
+// Stops when the context is cancelled.
+func (h *HealthChecker) StartReceiverPeriodicHealthCheck(ctx context.Context) {
+	for namespace, client := range h.clients {
+		go func(namespace string, client *azservicebus.Client) {
+			go h.periodicHealthCheck(ctx, h.receiverHealthCheck, namespace, client)
+		}(namespace, client)
+	}
+}
+
+func (h *HealthChecker) periodicHealthCheck(ctx context.Context, healthCheckFunc HealthCheckFunc, namespace string, client *azservicebus.Client) {
+	nextCheck := time.Now()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(time.Until(nextCheck)):
+			if err := healthCheckFunc(ctx, namespace, client); err != nil {
+				log(ctx, err)
+			}
+			nextCheck = nextCheck.Add(h.interval)
+		}
+	}
+}
+
+func (h *HealthChecker) senderHealthCheck(ctx context.Context, namespace string, client *azservicebus.Client) error {
+	s, err := client.NewSender(h.entity, nil)
+	if err != nil {
+		sender.Metric.IncHealthCheckFailureCount(namespace, h.entity)
+		return err
+	}
+	sbCtx, cancel := context.WithTimeout(ctx, h.options.HealthCheckTimeout)
+	defer cancel()
+	_, err = s.NewMessageBatch(sbCtx, nil)
+	if err != nil {
+		sender.Metric.IncHealthCheckFailureCount(namespace, h.entity)
+		return err
+	}
+	sender.Metric.IncHealthCheckSuccessCount(namespace, h.entity)
+	return s.Close(ctx)
+}
+
+func (h *HealthChecker) receiverHealthCheck(ctx context.Context, namespace string, client *azservicebus.Client) error {
+	var r *azservicebus.Receiver
+	var err error
+	if h.subscription == "" {
+		r, err = client.NewReceiverForQueue(h.entity, nil)
+	} else {
+		r, err = client.NewReceiverForSubscription(h.entity, h.subscription, nil)
+	}
+	if err != nil {
+		processor.Metric.IncHealthCheckFailureCount(namespace, h.entity, h.subscription)
+		return err
+	}
+	sbCtx, cancel := context.WithTimeout(ctx, h.options.HealthCheckTimeout)
+	defer cancel()
+	_, err = r.PeekMessages(sbCtx, 1, nil)
```

Comment:
does Peek blocks until it gets 1 message?
if there is no message, does it return an error?

Reply:
no message available doesn't return an error: https://github.com/Azure/azure-sdk-for-go/blob/d2f6a15a733221e831e16b7223366017e72b2875/sdk/messaging/azservicebus/internal/mgmt.go#L138-L141